File: backend/src/models/user.model.ts
import { Schema, model, Document, Types } from "mongoose"

export interface IAddress {
    _id?: Types.ObjectId
    recipient_name: string
    phone: string
    province_code: string
    district_code: string
    ward_code: string
    specific_address: string
    is_default: boolean
}

const AddressSchema = new Schema<IAddress>({
    recipient_name: { type: String, required: true, trim: true },
    phone: { type: String, required: true, trim: true },
    province_code: { type: String, required: true, trim: true },
    district_code: { type: String, required: true, trim: true },
    ward_code: { type: String, required: true, trim: true },
    specific_address: { type: String, required: true, trim: true },
    is_default: { type: Boolean, default: false }
}, { _id: true }) // cần id để chỉnh sửa

// Schemaa phương thuwc đnăg nhập
export interface IAuthProvider {
    provider: 'password' | 'google'
    provider_id?: string
}
const AuthProviderSchema = new Schema<IAuthProvider>({
    provider: { type: String, required: true, trim: true },
    provider_id: { type: String, required: true, trim: true }
}, { _id: false }) // k cần


export interface ILastLogin {
    device?: string | null
    ip?: string | null
    atTime?: Date | null
}

const LastLoginSchema = new Schema<ILastLogin>({
    device: { type: String, default: null, trim: true },
    ip: { type: String, default: null },
    atTime: { type: Date, default: null }
}, { _id: false }) // k cần bảng trung gian


export interface IUser extends Document {
    _id: Types.ObjectId
    email: string
    display_name: string
    password: string | null
    auth_provider?: Array<IAuthProvider>
    avatar_url?: string | null
    avatar_id?: string | null
    delivering_addresses: Types.DocumentArray<IAddress>
    roles: Array<"user" | "admin">
    is_active: boolean
    is_show: boolean
    last_login?: ILastLogin | null
    createdAt: Date
    updatedAt: Date
}


export const User = model<IUser>("users", new Schema<IUser>({
    email: {
        type: String,
        required: true,
        trim: true,
        lowercase: true,
        unique: true
    },
    display_name: {
        type: String,
        required: true,
        trim: true,
        index: true // để tìm kiếm nhanh trong admin
    },
    password: {
        type: String,
        default: null,  // cho phép null vì có thể qua social
        minlength: 8,
        select: false // không trả về khi find bằng id
    },
    auth_provider: {
        type: [AuthProviderSchema],
        default: null
    },
    avatar_url: { type: String, default: null, trim: true },
    avatar_id: { type: String, default: null, trim: true },
    delivering_addresses: {
        type: [AddressSchema],
        default: []
    },

    roles: {
        type: [String],
        enum: ["user", "admin"],
        default: ["user"]
    },

    is_active: { type: Boolean, default: true },
    is_show: { type: Boolean, default: false },

    last_login: { type: LastLoginSchema, default: null }
}, { timestamps: true }))

File: frontend_mobile/lib/features/auth/data/models/user_model.dart
// lib/features/auth/data/models/user_model.dart
class DeliveringAddress {
  final String id;
  final String recipientName;
  final String phone;
  final String provinceCode;
  final String districtCode;
  final String wardCode;
  final String specificAddress;
  final bool isDefault;

  DeliveringAddress({
    required this.id,
    required this.recipientName,
    required this.phone,
    required this.provinceCode,
    required this.districtCode,
    required this.wardCode,
    required this.specificAddress,
    required this.isDefault,
  });

  factory DeliveringAddress.fromJson(Map<String, dynamic> json) {
    return DeliveringAddress(
      id: json['_id'] as String? ?? '',
      recipientName: json['recipient_name'] as String? ?? '',
      phone: json['phone'] as String? ?? '',
      provinceCode: json['province_code'] as String? ?? '',
      districtCode: json['district_code'] as String? ?? '',
      wardCode: json['ward_code'] as String? ?? '',
      specificAddress: json['specific_address'] as String? ?? '',
      isDefault: json['is_default'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'recipient_name': recipientName,
      'phone': phone,
      'province_code': provinceCode,
      'district_code': districtCode,
      'ward_code': wardCode,
      'specific_address': specificAddress,
      'is_default': isDefault,
    };
  }
}

class LastLogin {
  final String device;
  final String? ip;
  final DateTime? atTime;

  LastLogin({required this.device, this.ip, this.atTime});

  factory LastLogin.fromJson(Map<String, dynamic> json) {
    return LastLogin(
      device: json['device'] as String? ?? '',
      ip: json['ip'] as String?,
      atTime: json['atTime'] != null
          ? DateTime.tryParse(json['atTime'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {'device': device, 'ip': ip, 'atTime': atTime?.toIso8601String()};
  }
}

class UserModel {
  final String id;
  final String email;
  final String? displayName;
  final String? authProvider;
  final String? avatarUrl;
  final String? avatarId;
  final List<String> roles;
  final bool isActive;
  final bool isShow;
  final LastLogin? lastLogin;
  final List<DeliveringAddress> deliveringAddresses;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  UserModel({
    required this.id,
    required this.email,
    this.displayName,
    this.authProvider,
    this.avatarUrl,
    this.avatarId,
    required this.roles,
    required this.isActive,
    required this.isShow,
    this.lastLogin,
    required this.deliveringAddresses,
    this.createdAt,
    this.updatedAt,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    final addressesJson =
        (json['delivering_addresses'] as List<dynamic>? ?? []);

    return UserModel(
      id: json['_id'] as String? ?? '',
      email: json['email'] as String? ?? '',
      displayName: json['display_name'] as String?,
      authProvider: json['auth_provider'] as String?,
      avatarUrl: json['avatar_url'] as String?,
      avatarId: json['avatar_id'] as String?,
      roles:
          (json['roles'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      isActive: json['is_active'] as bool? ?? true,
      isShow: json['is_show'] as bool? ?? true,
      lastLogin: json['last_login'] != null
          ? LastLogin.fromJson(json['last_login'] as Map<String, dynamic>)
          : null,
      deliveringAddresses: addressesJson
          .map((e) => DeliveringAddress.fromJson(e as Map<String, dynamic>))
          .toList(),
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'email': email,
      'display_name': displayName,
      'auth_provider': authProvider,
      'avatar_url': avatarUrl,
      'avatar_id': avatarId,
      'roles': roles,
      'is_active': isActive,
      'is_show': isShow,
      'last_login': lastLogin?.toJson(),
      'delivering_addresses': deliveringAddresses
          .map((e) => e.toJson())
          .toList(),
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }
}
